src/details/ArborX_DetailsVector.hpp

```diff
@@ -36,19 +37,33 @@ struct Vector
   KOKKOS_FUNCTION
   constexpr auto const &operator[](unsigned int i) const { return _coords[i]; }
 
-  KOKKOS_FUNCTION
-  constexpr auto dot(Vector const &w) const
+  template <typename Precision = Coordinate>
+  KOKKOS_FUNCTION constexpr auto dot(Vector const &w) const
   {
     auto const &v = *this;
 
-    Coordinate r = 0;
+    Precision r = 0;
     for (int d = 0; d < DIM; ++d)
-      r += v[d] * w[d];
+      r += static_cast<Precision>(v[d]) * static_cast<Precision>(w[d]);
     return r;
   }
 
-  KOKKOS_FUNCTION
-  auto norm() const { return Kokkos::sqrt(dot(*this)); }
+  template <typename Precision = Coordinate>
+  KOKKOS_FUNCTION auto norm() const
+  {
+    return Kokkos::sqrt(dot<Precision>(*this));
+  }
+
+  template <typename Precision = Coordinate>
+  KOKKOS_FUNCTION void normalize()
```

Comment:
Why do we want it as a non-const member function?

Reply:
I think there are few options:
```
(1)  void Vector::normalize();                 // normalize the vector
(2)  Vector Vector::normalize() const;  // return normalized vector
(3) void normalize(Vector&);                 // free function to normalize the vector
(4) Vector normalize(Vector const&);   // free function to return normalized vector
```
I don't have a strong preference for any. My approach was that typically if we normalize a vector, we don't need to keep the not normalized one. So, I think it would be preferable to modify the provided vector. Whether it should do (1) or (3) is less obvious.

Reply:
I lean towards (4) `normalize(V const&) -> V` or (4bis) `normalize(V&, V const&) -> void` that explicitly allows calling with 1st arg same as 2nd.

Reply:
I don't understand how this is better. Could you describe use cases or usage where it would be more convenient?

Reply:
Options | distinct args | in-place
--- | --- | ---
(1) | `auto w = v;`<br> `w.normalize();` | `v.normalize();`
(2) | `auto w = v.normalize();` | `v = v.normalize();`
(3) | `auto w = v;`<br> `normalize(w);` | `normalize(v);`
(4) | `auto w = normalize(v);` | `v = normalize(v);`
(4bis) | `Vector w;`<br> `normalize(w, v);` | `normalize(v, v);`

I find (1) awkward.  I also note that this is the only member function that modifies the underlying geometric entity.

Reply:
Thanks for writing these down. When I see all of them written like this, I think I agree that (4) is probably the most easy to understand one. The only downside is potentially creating a copy for in-place.

Reply:
I'm not sure (4) is doable in terms of specifying precision. We would need something like
```c++
template<int DIM, typename Coordinate, typename Precision = Coordinate>
auto normalize(Vector<DIM, Coordinate> const& v);
```
However, this won't work with
```
auto w = normalize<double>(v);
```
as the template argument order is wrong. I could be missing something.

Reply:
@masterleinad helped me with the previous comment. I opted to do (4) for now, as doing `normalize(v, v)` still seems wrong to me.